# Overview

This notebook assumes you have already collected and scored MOC tweets. It creates a dataset for use in R to analyze the patterns of polarization over time. You will do some parsing on an AWS server and some locally before ultimately making a CSV file that you can open and analyze in R.

LH Note: on my computer, Git stuff and data live in different places, so you'll see notes about moving files or changing directories. I haven't figured out a good way to keep files in both places or to mirror or sync or whatever. So, for now, paths are hard-coded or there's a note about where to find a file.

# Get Data

2016 election data is on an AWS server under ```/data/purpletag```. 

To login: 

```ssh -i ~/.ssh/carolgrrr.pem ubuntu@purpletag.casmlab.org```

The data is large (> 4GB), so best to run Juypter notebooks to parse on the server. Then CSV files can be used locally.

You can run a notebook on the server and use your local browser with these two commands:

* ```ssh -L 8080:localhost:8888 -i ~/.ssh/carolgrrr.pem ubuntu@purpletag.casmlab.org```
* ```nohup jupyter notebook --no-browser > log.txt 2>&1 &```

Then access ```http://localhost:8080``` in your browser.

## On server: Parsing from ```scores``` files to CSV

This section assumes you have already run purpletag's ```collect``` and ```score``` functions and gotten the Twitter data that you want in JSON format and parsed that data into score files.

In [ ]:
# Based on https://stackoverflow.com/questions/26415906/read-multiple-txt-files-into-pandas-dataframe-with-filename-as-column-header
import pandas as pd
import os
import glob

# Step 1: get a list of all score files in target directory
my_dir = "/data/purpletag/scores/"
filelist = []
filesList = []
os.chdir( my_dir )

# Step 2: Build up list of files:
for files in glob.glob("*.1.moc.scores"): # using 1-day purpletag MOC scores
    fileName, fileExtension = os.path.splitext(files)
    filesList.append(files) #filename with extension

In [ ]:
# Step 3: Build up DataFrame:
# Based on https://stackoverflow.com/questions/35717706/python-how-to-turn-a-dictionary-of-dataframes-into-one-big-dataframe-with-colum
d = {} # dictionary to hold multiple dfs

for filename in filesList:
    df1 = pd.read_csv(filename, header=None, sep=' ', index_col=0)
    d[filename[:-13]] = df1
    
df = pd.concat(d, axis=1)
df.columns = df.columns.droplevel(-1) 

df.to_pickle('/data/purpletag/mocs_by_date.pkl')

Move the file from the AWS server to local if you want to work locally:

```scp -i ~/.ssh/carolgrrr.pem ubuntu@purpletag.casmlab.org:/data/purpletag/mocs_by_date.pkl ~/Documents/git/casmlab/purpletag/files/```

## Locally: Prepping for stats

We now have a pickled dataframe of the form handleXdate. We need to keep data only from Labor Day to Election Day and get weekly averages.

In [146]:
import pandas as pd

df = pd.read_pickle('data-files/mocs_by_date.pkl')
df.head()

2015-11-10  2015-11-11  2015-11-12  2015-11-13  2015-11-14  \
austinscottga08         NaN     2.40933         NaN         NaN         NaN   
benniegthompson         NaN         NaN         NaN         NaN         NaN   
bettymccollum04         NaN         NaN         NaN         NaN         NaN   
billpascrell            NaN         NaN   -1.168750     -0.3731   -0.972477   
boblatta                NaN         NaN    0.965309         NaN         NaN   

                 2015-11-15  2015-11-16  2015-11-17  2015-11-18  2015-11-19  \
austinscottga08         NaN         NaN         NaN         NaN         NaN   
benniegthompson         NaN         NaN     7.56379         NaN         NaN   
bettymccollum04         NaN         NaN         NaN         NaN         NaN   
billpascrell            NaN         NaN         NaN         NaN         NaN   
boblatta           0.612764         NaN     1.56897         NaN   -0.031978   

                    ...      2016-10-30  2016-10-31  2016-11-01  2016-11-02  \
austinscottga08     ...             NaN         NaN         NaN         NaN   
benniegthompson     ...             NaN         NaN         NaN         NaN   
bettymccollum04     ...             NaN         NaN    -90.6103    -69.5725   
billpascrell        ...             NaN         NaN         NaN    -18.7023   
boblatta            ...             NaN         NaN         NaN         NaN   

                 2016-11-03  2016-11-04  2016-11-05  2016-11-06  2016-11-07  \
austinscottga08         NaN         NaN         NaN         NaN         NaN   
benniegthompson         NaN         NaN         NaN         NaN         NaN   
bettymccollum04    -50.1429         NaN    -33.4513         NaN         NaN   
billpascrell            NaN         NaN         NaN         NaN         NaN   
boblatta                NaN         NaN         NaN         NaN         NaN   

                 2016-11-08  
austinscottga08         NaN  
benniegthompson         NaN  
bettymccollum04         NaN  
billpascrell            NaN  
boblatta                NaN  

[5 rows x 365 columns]

In [147]:
week1_dates = ['2016-09-06','2016-09-07','2016-09-08','2016-09-09','2016-09-10','2016-09-11','2016-09-12']
week2_dates = ['2016-09-13','2016-09-14','2016-09-15','2016-09-16','2016-09-17','2016-09-18','2016-09-19']
week3_dates = ['2016-09-20','2016-09-21','2016-09-22','2016-09-23','2016-09-24','2016-09-25','2016-09-26']
week4_dates = ['2016-09-27','2016-09-28','2016-09-29','2016-09-30','2016-10-01','2016-10-02','2016-10-03']
week5_dates = ['2016-10-04','2016-10-05','2016-10-06','2016-10-07','2016-10-08','2016-10-09','2016-10-10']
week6_dates = ['2016-10-11','2016-10-12','2016-10-13','2016-10-14','2016-10-15','2016-10-16','2016-10-17']
week7_dates = ['2016-10-18','2016-10-19','2016-10-20','2016-10-21','2016-10-22','2016-10-23','2016-10-24']
week8_dates = ['2016-10-25','2016-10-26','2016-10-27','2016-10-28','2016-10-29','2016-10-30','2016-10-31']
week9_dates = ['2016-11-01','2016-11-02','2016-11-03','2016-11-04','2016-11-05','2016-11-06','2016-11-07']

df['week1'] = df[week1_dates].mean(axis=1)
df['week2'] = df[week2_dates].mean(axis=1)
df['week3'] = df[week3_dates].mean(axis=1)
df['week4'] = df[week4_dates].mean(axis=1)
df['week5'] = df[week5_dates].mean(axis=1)
df['week6'] = df[week6_dates].mean(axis=1)
df['week7'] = df[week7_dates].mean(axis=1)
df['week8'] = df[week8_dates].mean(axis=1)
df['week9'] = df[week9_dates].mean(axis=1)

df.head()

2015-11-10  2015-11-11  2015-11-12  2015-11-13  2015-11-14  \
austinscottga08         NaN     2.40933         NaN         NaN         NaN   
benniegthompson         NaN         NaN         NaN         NaN         NaN   
bettymccollum04         NaN         NaN         NaN         NaN         NaN   
billpascrell            NaN         NaN   -1.168750     -0.3731   -0.972477   
boblatta                NaN         NaN    0.965309         NaN         NaN   

                 2015-11-15  2015-11-16  2015-11-17  2015-11-18  2015-11-19  \
austinscottga08         NaN         NaN         NaN         NaN         NaN   
benniegthompson         NaN         NaN     7.56379         NaN         NaN   
bettymccollum04         NaN         NaN         NaN         NaN         NaN   
billpascrell            NaN         NaN         NaN         NaN         NaN   
boblatta           0.612764         NaN     1.56897         NaN   -0.031978   

                   ...     2016-11-08       week1       week2      week3  \
austinscottga08    ...            NaN         NaN    4.932035        NaN   
benniegthompson    ...            NaN         NaN   -1.211380        NaN   
bettymccollum04    ...            NaN -120.131512 -467.500940 -89.054550   
billpascrell       ...            NaN   -2.871430   -2.577447  -2.353348   
boblatta           ...            NaN    1.468710   13.197433        NaN   

                     week4      week5      week6     week7      week8  \
austinscottga08   1.064290  14.208100   6.042600  1.561610        NaN   
benniegthompson        NaN        NaN  -1.823010       NaN        NaN   
bettymccollum04 -86.397562 -29.160648 -35.474095 -5.049413 -33.368257   
billpascrell     -3.437217        NaN  -1.026550       NaN  -1.333330   
boblatta               NaN        NaN   0.974138       NaN        NaN   

                    week9  
austinscottga08       NaN  
benniegthompson       NaN  
bettymccollum04 -60.94425  
billpascrell    -18.70230  
boblatta              NaN  

[5 rows x 374 columns]

In [148]:
weekly_df = df[['week1','week2','week3','week4','week5','week6','week7','week8','week9']]
weekly_df

week1       week2       week3       week4       week5  \
austinscottga08         NaN    4.932035         NaN    1.064290   14.208100   
benniegthompson         NaN   -1.211380         NaN         NaN         NaN   
bettymccollum04 -120.131512 -467.500940  -89.054550  -86.397562  -29.160648   
billpascrell      -2.871430   -2.577447   -2.353348   -3.437217         NaN   
boblatta           1.468710   13.197433         NaN         NaN         NaN   
bradsherman             NaN   -1.941180         NaN  -10.763900         NaN   
call_me_dutch     -2.754159   -5.169143   -3.129013   -1.573270   -2.967365   
candicemiller      1.483315         NaN   -0.137418   -0.264974         NaN   
cathymcmorris     31.845395    4.472548   29.771150   39.539832    1.000000   
cbrangel         -59.337715 -143.615365  -80.236928  -66.473658  -19.118496   
chelliepingree    -1.914142   -1.050367   -0.218658   -4.608034   -1.159367   
chrismurphyct    -21.378306 -114.012478  -14.055037   -8.313174  -10.745990   
chrisvanhollen  -139.429865 -666.581665  -16.767200         NaN  -16.172200   
chuckgrassley           NaN    6.196710   19.649100    5.808580    0.906780   
clyburn           -0.957143 -392.801880  -56.408405  -62.113000         NaN   
conawaytx11             NaN         NaN         NaN         NaN         NaN   
congbillposey      0.967105         NaN         NaN         NaN         NaN   
congculberson      2.081748   28.792382         NaN   14.199485    0.906780   
conghuelskamp      3.900722   10.150968   48.922836   17.092080   22.739926   
congmikesimpson    2.104140    0.757014    1.926503    1.064290    3.609000   
congpalazzo        2.823690         NaN   -1.052340   17.634900         NaN   
congressmangt      6.994534   21.155100   24.857698    6.832215   12.747636   
congressmanhice    5.573510   32.364031    3.812027    4.799283    5.160796   
congressmanruiz   -0.964939         NaN         NaN   -2.495040   -0.966387   
corybooker      -144.405620   -8.007561  -14.649831  -95.775025  -12.866565   
cynthialummis      0.982223         NaN    6.895440         NaN   38.926490   
darrellissa             NaN   12.797176    1.780650         NaN         NaN   
daveloebsack       1.468710    0.507460   -1.222368   -7.857877         NaN   
davereichert      -0.124652   10.309078    5.510810    3.508307    5.389447   
desjarlaistn04     0.904247    8.660580    2.341166         NaN    3.175186   
...                     ...         ...         ...         ...         ...   
sentomcotton      49.562390   39.572987    9.303593    2.846112    2.585690   
sentoomey         -0.705847    1.409045         NaN    1.682178         NaN   
senwarren       -185.152961 -153.356397         NaN  -17.836100         NaN   
senwhitehouse    -59.135102  -52.270855   -7.241852  -53.877967   -2.246048   
speakerryan      425.967749  283.880371  399.274100  380.479214  265.460197   
staceyplaskett   -93.786500  -72.971800  -46.655300   -6.911350  -14.704901   
stevedaines       -0.170952   24.060960    7.654405    3.611700    1.992373   
stevekingia        0.884615         NaN         NaN    1.058930         NaN   
steveknight25      0.011508         NaN         NaN    1.259380         NaN   
stevescalise      23.487002   49.402728   24.582102   23.764230   45.489375   
susanwbrooks      -3.005262   54.109362    7.574565    7.944950    0.906780   
teammoulton      -29.967346 -192.858030  -20.186438  -10.653550  -22.170407   
tgowdysc           1.468710   82.782623   11.402900    7.369750         NaN   
toddrokita        13.166180    9.046467    6.462324   23.332543    2.253390   
tomcoleok04        0.186740    2.870305    0.890323         NaN         NaN   
tomrooney               NaN    4.191300         NaN         NaN         NaN   
tulsipress        -7.338243  -16.907528   -7.251179  -16.793391   -5.940548   
txrandy14         -4.429010    5.573515    2.618932   12.821244         NaN   
usrepdavidjolly  -59.315860  -21.616927         NaN    2.528815    1.843

In [164]:
import pandas as pd
import yaml

# get the data from Govtrack
with open('/Users/libbyh/Dropbox/CASM/SMCE/Shared Social Media and Civic Engagement/Data/purpletag/legislators-social-media.yaml', 'r') as f:
    df_social = pd.io.json.json_normalize(yaml.load(f))

with open('/Users/libbyh/Dropbox/CASM/SMCE/Shared Social Media and Civic Engagement/Data/purpletag/legislators-current.yaml', 'r') as f:
    df_current = pd.io.json.json_normalize(yaml.load(f))

print(len(weekly_df))
# merge everything into one data frame with one row per MOC
df_meta = pd.merge(df_current, df_social, on="id.govtrack")
df_meta["handle"] = df_meta["social.twitter"].str.lower()
weekly_df["handle"] = weekly_df.index.str.lower()

print(len(df_meta))

df_merged = pd.merge(df_meta, weekly_df, left_on="handle", right_index=True)

print(len(df_merged))

#cols_to_keep = ['id.govtrack','social.twitter','name.official_full','bio.gender','terms','week1','week2','week3','week4','week5','week6','week7','week8','week9']

df_merged = df_merged[['id.govtrack','social.twitter','name.official_full','bio.gender','terms','week1','week2','week3','week4','week5','week6','week7','week8','week9']]


511
529
444


Not sure why we have only 444 matches, but it's better than 12.

In [165]:
df1 = pd.concat([df_merged.drop(['terms'], axis=1), df_merged['terms'].apply(pd.Series)], axis=1)
df2 = pd.concat([df1.drop([0], axis=1), df1[0].apply(pd.Series)], axis=1)

keep_df = df2[['id.govtrack','social.twitter','name.official_full','bio.gender','type','party','week1','week2','week3','week4','week5','week6','week7','week8','week9']]
keep_df

id.govtrack   social.twitter      name.official_full bio.gender type  \
0         400050  SenSherrodBrown           Sherrod Brown          M  rep   
1         300018  SenatorCantwell          Maria Cantwell          F  rep   
2         400064    SenatorCardin      Benjamin L. Cardin          M  rep   
3         300019    SenatorCarper        Thomas R. Carper          M  rep   
4         412246      SenBobCasey    Robert P. Casey, Jr.          M  sen   
5         412248     SenBobCorker              Bob Corker          M  sen   
6         300043     SenFeinstein        Dianne Feinstein          F  sen   
7         300052    SenOrrinHatch          Orrin G. Hatch          M  sen   
9         412243  McCaskillOffice        Claire McCaskill          F  sen   
10        400272  SenatorMenendez         Robert Menendez          M  rep   
11        300078    SenBillNelson             Bill Nelson          M  rep   
13        300093      SenStabenow         Debbie Stabenow          F  rep   
14        412244    SenatorTester              Jon Tester          M  sen   
15        412247    SenWhitehouse      Sheldon Whitehouse          M  sen   
16        412251  SenJohnBarrasso           John Barrasso          M  sen   
17        400432    SenatorWicker         Roger F. Wicker          M  rep   
18        300002     SenAlexander         Lamar Alexander          M  sen   
19        300023   SenThadCochran            Thad Cochran          M  rep   
20        300025   SenatorCollins        Susan M. Collins          F  sen   
21        300027       JohnCornyn             John Cornyn          M  sen   
22        300038    SenatorDurbin       Richard J. Durbin          M  rep   
23        300041      SenatorEnzi         Michael B. Enzi          M  sen   
24        300047       GrahamBlog          Lindsey Graham          M  rep   
25        300055      InhofePress         James M. Inhofe          M  rep   
26        300072   McConnellPress         Mitch McConnell          M  sen   
27        412325   SenJeffMerkley            Jeff Merkley          M  sen   
28        300081      SenJackReed               Jack Reed          M  rep   
29        412322     SenatorRisch          James E. Risch          M  sen   
30        300083    SenPatRoberts             Pat Roberts          M  rep   
31        412323   SenatorShaheen          Jeanne Shaheen          F  sen   
..           ...              ...                     ...        ...  ...   
438       412641   RepDavidRouzer            David Rouzer          M  rep   
439       412643  RepTomMacArthur        Thomas MacArthur          M  rep   
440       412644        RepBonnie   Bonnie Watson Coleman          F  rep   
441       412646     RepLeeZeldin           Lee M. Zeldin          M  rep   
442       412647  RepKathleenRice        Kathleen M. Rice          F  rep   
443       412648      RepStefanik       Elise M. Stefanik          F  rep   
444       412649     RepJohnKatko              John Katko          M  rep   
445       412650       RepRussell           Steve Russell          M  rep   
446       412651  RepRyanCostello        Ryan A. Costello          M  rep   
448       412653     RepRatcliffe          John Ratcliffe          M  rep   
449       412654    hurdonthehill               Will Hurd          M  rep   
450       412655    RepBrianBabin             Brian Babin          M  rep   
451       412656       repmialove             Mia B. Love          F  rep   
452       412657      repdonbeyer    Donald S. Beyer, Jr.          M  rep   
453       412658      RepComstock        Barbara Comstock          F  rep   
454       412659   staceyplaskett      Stacey E. Plaskett          F  rep   
455       412660      RepNewhouse            Dan Newhouse          M  rep   
456       412661      RepGrothman          Glenn Grothman          M  rep   
457       412662    RepAlexMooney     Alexander X. Mooney          M  rep   
458       412663   RepEvanJenkins         Evan H. Jenkins          M  rep   
4

In [170]:
# melt it so each row is a person x week
df_long = pd.melt(keep_df, id_vars=['id.govtrack','social.twitter','name.official_full','bio.gender','party','type'],
                value_vars=['week1','week2','week3','week4','week5','week6','week7','week8','week9'],
                var_name='week', value_name='avg_score')
df_long['week'] = df_long['week'].str[-1:]

df_long.rename(columns = {'type':'chamber', 'social.twitter': 'handle', 'name.official_full': 'name', 'bio.gender': 'gender'}, inplace = True)

df_long.head()

id.govtrack           handle                  name gender     party  \
0       400050  SenSherrodBrown         Sherrod Brown      M  Democrat   
1       300018  SenatorCantwell        Maria Cantwell      F  Democrat   
2       400064    SenatorCardin    Benjamin L. Cardin      M  Democrat   
3       300019    SenatorCarper      Thomas R. Carper      M  Democrat   
4       412246      SenBobCasey  Robert P. Casey, Jr.      M  Democrat   

  chamber week   avg_score  
0     rep    1   -0.222430  
1     rep    1 -118.229375  
2     rep    1  -25.636298  
3     rep    1 -155.209127  
4     sen    1  -36.255503

In [175]:
# make sure we have just two parties
df_long.party.unique()

array(['Democrat', 'Republican'], dtype=object)

In [176]:
# get an absolute value of the polar score
df_long['abs'] = df_long['avg_score'].abs()

In [177]:
df_long.to_csv('data-files/weekly_averages_long.csv')

# Now move to R for analysis

Run ```~/Documents/git/casmlab/purpletag/2016_election.R```

That R script sends its output to ```2016_election_results.txt```

In [180]:
results = open('data-files/2016_election_results.txt', 'r')
print(results.read())


> # for pretty regression tables
> # http://stackoverflow.com/questions/30195718/stargazer-save-to-file-dont-show-in-console
> mod_stargazer <- functi .... [TRUNCATED] 

> df <- read.csv('weekly_averages_long.csv', header = TRUE, sep = ",", quote = "\"",
+                dec = ".", fill = TRUE, comment.char = "")

> summary(df)
       X           id.govtrack                 handle                 name      gender  
 Min.   :   0.0   Min.   :300002   AustinScottGA08:   9   Adam B. Schiff:   9   F: 792  
 1st Qu.: 998.8   1st Qu.:400326   BennieGThompson:   9   Adam Kinzinger:   9   M:3204  
 Median :1997.5   Median :412292   BettyMcCollum04:   9   Adam Smith    :   9           
 Mean   :1997.5   Mean   :401868   BillPascrell   :   9   Adrian Smith  :   9           
 3rd Qu.:2996.2   3rd Qu.:412533   BobLatta       :   9   Al Franken    :   9           
 Max.   :3995.0   Max.   :412674   BradSherman    :   9   Al Green      :   9           
                                   (Other)    

Based on the outlier-excluded linear mixed-effects models, it makes sense to remove RepThompson. The pattern stays the same even with RepThompson in the set though: negative effect for republican and week, positive effect for their interaction. ```lmm5``` is the model-of-best-fit. 

## Changing the way we score hashtags

What if we score tags for the 63-day period and then score MOCS?

Run the following (on the server) to get new scores:

* purpletag parse -t 63 -d 200
* purpletag score
* purpletag score --counts --score-mocs

That first command took a week because the code starts with today and works backwards 200 days, one day at a time. Each day takes over an hour. See Issue #18 about options for changing this behavior.

With the new tag measures, can start the process over. Start at "On Server: Parsing scores to CSV" with a new file name.